File: code/ch10/10.2.3-house-rob.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        memo = [0 for x in range(len(nums) + 1)]
        memo[-2] = nums[-1]

        for i in range(len(nums) - 2, -1, -1):
            memo[i] = max(memo[i + 1], memo[i + 2] + nums[i])

        return memo[0]

File: code/ch10/10.2.3-house-rob.java
class Solution {

    public int rob(int[] nums) {
        if(nums.length <= 0) return 0;

        int[] memo = new int[nums.length+1];
        memo[nums.length-1] = nums[nums.length-1];

        for (int i = nums.length-2; i >= 0; i--) {
            memo[i] = Math.max(memo[i+1], memo[i+2]+nums[i]);
        }
        return memo[0];
    }
}